### PR TITLE
Fix realtime concurrency: close connections on worker exit, fold over a trailing window

### PR DESCRIPTION
**Pull request by lohanidamodar**
## Summary

Two changes that together stop the `realtime.connections` concurrency gauge from ratcheting upward — the cause of the realtime peak rising 3.7× in five days on Cloud billing.

1. **Close a worker's connections when it exits** (`app/realtime.php`, new `Appwrite\Utopia\WebSocket\Adapter\Swoole`) — the root cause.
2. **Fold the level over a trailing window instead of carrying it** (`Appwrite\Usage\Concurrency`) — bounds any residual error and removes the manual re-anchor.

## 1. Why closes were lost

In Swoole process mode the connections belong to the reactor, while the registry that maps a connection to its project (`$realtime->connections`) lives in the worker that accepted it. When that worker exits — rollout, HPA scale-down, reload — its connections stay open. Whenever they finally close, `onClose` runs in a worker that never registered them, the `array_key_exists` gate fails, and the `-1` is never emitted. Every worker exit therefore added its open count to the level permanently.

The fix hooks `WorkerExit`, which fires in the exiting worker **while its event loop is still alive** (unlike `onWorkerStop`), and closes every registered connection with `1001 Going Away`. The `-1`, the `connectionsTotal` decrement, telemetry, presence cleanup and `unsubscribe` all take the existing `onClose` path — no second bookkeeping, no double count — and clients reconnect to a live worker instead of holding a socket the worker has abandoned. The base Utopia adapter exposes no such hook, so a small subclass adds it and enables `reload_async` with `max_wait_time` 15s, which is what lets the loop drain. Whoever stops the process caps that anyway -- a Cloud pod leaves ~10s after its 20s preStop, out of a 30s grace period -- so the value is set high enough not to be the tighter limit.

## 2. Why the fold also changes

Even with the flush, a killed worker (OOM, `max_wait_time` exceeded) still loses closes, and a carried level keeps every such error forever. Each bucket is now

```
level = max(0, Σ deltas in (bucket end − window, bucket end])
```

with a **6h** window (`Concurrency::WINDOW_SECONDS`, a constant like the interval and the lag). No carried state, so an orphaned open is forgotten one window later, at the price of not seeing connections older than the window.

A tenant is still written only for the buckets it had deltas in — the window changes the value, not how often it is sampled — so gauge write volume is unchanged. The event read now spans the window as well as the catch-up range; because it is time-ascending, a result that hits the 50,000-row cap is still complete for every bucket before the last one returned, so the fold samples up to there and the next run resumes, rather than writing levels missing deltas and advancing past them.

### Calibration (production, read-only, 2026-09-08 02:20 UTC)

Live inbound connections counted on the realtime pods against each estimator for the same instant, ~3h after a fleet-wide rollout:

| region | live | 3h | **6h** | 12h | 24h | carried (gauge today) |
|---|---|---|---|---|---|---|
| fra1 | 1,415 | 461 | **1,142** | 2,519 | 3,957 | 29,521 |
| nyc3 | 1,263 | 1,482 | **2,499** | 5,617 | 11,298 | 35,238 |
| sgp1 | 300 | 188 | **314** | 662 | 828 | 4,180 |
| sfo3 / tor1 / syd1 | 161 / 73 / 63 | — | 52 / 28 / 33 | 144 / 77 / 51 | — | 2,692 / 758 / 207 |
| **total** | **3,275** | | | | | **72,596** |

Shorter windows hide long-lived connections (fra1 at 3h: 0.33×); longer ones carry a stopped worker's opens (nyc3 at 12h: 4.4×). 6h is the balance **with the leak still present**. Once (1) is confirmed in production the leak term disappears and a longer window (24h+) becomes the better default — deliberately not changed here until that is verified.

## Behaviour that does not change

- Same gauge, metric name and 5-minute cadence; resume-from-last-sample unchanged, so no bucket is emitted twice.
- Cloud's `getRealtimePeakConnections` (max over the cycle's samples) and `/v1/usage/gauges?aggregate=max` read exactly as before.
- Gauge write volume: a tenant is sampled only for buckets it had deltas in, as before.

## Tests and gaps

`tests/unit/Usage/ConcurrencyTest.php` — each case seen red against the carried fold, green after:
- the level is the windowed sum, not the carried gauge
- an open exactly one window old has aged out
- each catch-up bucket gets its own window
- a bucket without deltas is not sampled
- a capped read samples only the buckets it completed
- the event read starts one window before the first emitted bucket

**No automated test covers the worker-exit path** — it is the Swoole server lifecycle. It was instead verified against a real stack (compose, ClickHouse, the stats-usage worker), opening websocket connections and restarting the realtime container with SIGTERM, which is the pod-termination path:

| realtime.php under test | opens | closes | net |
|---|---|---|---|
| `origin/main` | 8 | 0 | **8 stranded** |
| this branch | 10 | 10 | **0** |
| adapter only (`reload_async`, no exit hook) | 8 | 0 | 8 stranded |
| exit hook registered but empty | 8 | 0 | 8 stranded |

The last two rows matter: neither enabling `reload_async` nor merely registering a `workerExit` handler is sufficient. The closes come from the hook's own sweep, which drives them through the normal `onClose` path.

Running it also turned up a defect that no unit test would have: Swoole re-invokes `onWorkerExit` until the worker's loop is empty, so the sweep ran on every call — a few hundred warnings per shutdown and repeated `close()` on connections already closing. It now sweeps once; a shutdown writes exactly one line, `Worker 0 exiting, closing 10 open connections`.

Still worth watching on the first production rollout: whether a worker holding ~470 connections drains inside the ~10s a pod has after its preStop. Each close does database and queue work, and the local run drained only a dozen. The log line gives the count to compare against the closes that land that hour; if it falls short the remainder is stranded, though the 6h window now caps how long that costs anything.

## Follow-ups (Cloud)

- Bump `REALTIME_CONCURRENCY_BASELINE` to the deploy time so this cycle's peak is taken only over windowed samples.
- Refresh the `getRealtimePeakConnections` docblock, which still describes the carried fold.
- Raise `WINDOW_SECONDS` once the flush is confirmed in production.

### CODE REVIEW

File: src/Appwrite/Usage/Concurrency.php [hunk 1]
```diff
@@ -41,54 +39,87 @@ class Concurrency
...
+
         $rows = $usage->findAcrossTenants([
             UsageFilter::equal('metric', [METRIC_REALTIME_CONNECTIONS]),
-            UsageFilter::greaterThanEqual('time', $start->format('Y-m-d H:i:s')),
+            UsageFilter::greaterThanEqual('time', $readFrom->format('Y-m-d H:i:s')),
             UsageFilter::lessThan('time', $end->format('Y-m-d H:i:s')),
             UsageQuery::groupBy('tenant'),
             UsageQuery::groupByInterval('time', REALTIME_CONCURRENCY_INTERVAL),
             UsageFilter::orderAsc('time'),
             UsageFilter::limit(self::MAX_ROWS),
```
[the remark this answers is not in the document]
Comment (lohanidamodar) [reply src/Appwrite/Usage/Concurrency.php]: Valid. The read is time-ascending, so a capped result is still complete for every bucket before the last one it returned. In ea7c4a0cb6 the fold samples only up to that bucket and the next run resumes from there; if a single bucket alone exceeds the cap it logs a warning and writes nothing rather than a partial level. Covered by `testCappedReadSamplesOnlyTheBucketsItCompleted`, seen red against the previous fold.

File: src/Appwrite/Usage/Concurrency.php [hunk 2]
```diff
@@ -41,54 +39,87 @@ class Concurrency
...
+                }
+
+                // Negative when the closes we see belong to opens we no longer do.
+                $samples[] = [
+                    'tenant' => $tenant,
+                    'metric' => METRIC_REALTIME_CONNECTIONS,
+                    'value' => \max(0, $level),
+                    'time' => (new \DateTime('@' . $at))->setTimezone(new \DateTimeZone('UTC')),
+                ];
+            }
```
[the remark this answers is not in the document]
Comment (lohanidamodar) [reply src/Appwrite/Usage/Concurrency.php:122]: Valid — that loop wrote every catch-up bucket for any tenant active in the window. ea7c4a0cb6 reverts to writing a tenant only for the buckets it had deltas in, so the volume is what it was; the window changes the value, not how often it is sampled. `testBucketWithoutDeltasIsNotSampled` pins it.

File: src/Appwrite/Usage/Concurrency.php [hunk 3]
```diff
@@ -101,30 +132,14 @@ public function sample(Usage $usage): int
...
-        $levels = [];
-        foreach ($rows as $row) {
-            $tenant = $row->getTenant();
-            if ($tenant !== '' && $tenant !== null) {
-                $levels[$tenant] = (int) $row->getValue();
-            }
-        }
-
-        return $levels;
+        return \max(self::INTERVAL_SECONDS, (int) System::getEnv('_APP_REALTIME_CONCURRENCY_WINDOW', 21600));
```
[the remark this answers is not in the document]
Comment (lohanidamodar) [reply src/Appwrite/Usage/Concurrency.php]: Added in ea7c4a0cb6 to `app/config/variables.php`, `.env` and the three `docker-compose.yml` services that carry `_APP_STATS_RESOURCES_INTERVAL`; seconds, default 21600.

[the remark this answers is not in the document]
Comment (lohanidamodar) [reply src/Appwrite/Usage/Concurrency.php]: Superseded in d0c4770287: the window is no longer an environment variable. Only the fold reads it, so it is a class constant (`Concurrency::WINDOW_SECONDS`, 6h) like the interval and the lag, and the `variables.php` / `.env` / compose additions are reverted — nothing to wire.

File: src/Appwrite/Usage/Concurrency.php [hunk 4]
```diff
@@ -27,68 +29,115 @@ class Concurrency
...
+        if (\count($rows) >= self::MAX_ROWS) {
+            $lastTime = new \DateTime((string) \end($rows)->getAttribute('time'), new \DateTimeZone('UTC'));
+            $cut = \intdiv($lastTime->getTimestamp(), self::INTERVAL_SECONDS) * self::INTERVAL_SECONDS;
+
+            if ($cut <= $start->getTimestamp()) {
+                Console::warning('Realtime concurrency: over ' . self::MAX_ROWS . ' tenant buckets before the first bucket closed; nothing sampled');
+
+                return 0;
+            }
+
```
[the remark this answers is not in the document]
Comment (lohanidamodar) [reply src/Appwrite/Usage/Concurrency.php]: Valid, and worse than a truncation: the resume point comes from the newest gauge sample, so when the cap lands at or before the first bucket awaiting a sample nothing is written, the start time never moves, and every later run repeats the same capped read.

Fixed in b8ca0ee369 by paging the read rather than stopping at the cap. Each page drops its newest bucket, which the limit may have split, and the next page re-reads that bucket whole, so the cursor advances by at least one interval per page. Only a bucket that fills a page on its own is unsamplable, and that is warned about rather than silently truncated. Page size is a constructor argument so a test can reach the cap without 50,000 rows.

Two tests cover it, both seen red against the previous commit (with its cap shrunk so it is reachable): `testDenseHistoryBeforeTheFirstBucketDoesNotStall` is exactly the scenario you describe, and `testDenseReadIsPagedUntilExhausted` pins that a bucket split across a page boundary is counted once and in full.

File: tests/unit/Usage/ConcurrencyTest.php [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+        foreach ($eventReads[0][1] as $query) {
+            if ($query->getAttribute() === 'time' && \in_array($query->getMethod(), [Method::GreaterThanEqual, Method::LessThan], true)) {
+                $bounds[$query->getMethod()->value] = $query->getValue();
+            }
+        }
+
+        $this->assertSame([
+            Method::GreaterThanEqual->value => '2026-09-08 06:00:00',
+            Method::LessThan->value => '2026-09-08 12:00:00',
+        ], $bounds);
```
[the remark this answers is not in the document]
Comment (lohanidamodar) [reply tests/unit/Usage/ConcurrencyTest.php]: Agreed, and it matches the repo's own rule about asserting observable behaviour. Fixed in b06189ded4.

`testEventReadCoversOneWindowBeforeTheFirstEmittedBucket` is gone and so is the event-read count in the paging test; the recorded-reads log they needed went with them. The boundary they guarded is already observable in the sampled output: in `testLeakedOpensAgeOutOfTheWindow` an open exactly one window back has aged out of the level, and in `testLevelIsTheWindowedSumNotTheCarriedGauge` one inside the window still counts toward it. If the read stopped covering a full window before the first bucket, both would fail on their values.

On the fixture sized from `MAX_ROWS` — that is already gone too; the page size is a constructor argument now, so the cap is reachable with three rows instead of fifty thousand.

Coverage is unchanged, checked rather than assumed: with the clock pinned, the carried fold fails six of the seven remaining tests, and the pre-paging read fails the dense-paging one.
